### ISSUE TITLE
[Proposal] List Comprehension in function definitions

### ISSUE THREAD

**Issue by DruidGreeneyes** [cl21/cl21#85]
I'd like to have a reader, which might be `l(...)` or `$(...)` or `?(...)` or something else, but which will allow this: `(defun my-fun ($(foo . bar)) ...)` to be equivalent to this:
```
(defun my-fun (foo)
  (let ((bar (car foo)) (baz (cdr foo))) 
    ...)
```
Or this: `(defun my-fun ($(_ . foo)) ..)` to be equivalent to this:
```
(defun my-fun (foo)
  (let ((bar (cdr foo)))
    ...)
```
Ideally I'd like to be able to extend the construct to offer sophisticated sequence-destructuring in the parameter-set of a function, so that something like `$(_ . (x . _)))` might take the `cdar` of a list or the first element of the second nested array. 

Something like that, anyway. I'd be happy with the first two. The last is just icing.

Anyway, I love this idea. Keep it up.

**Comment by guicho271828** [comment]
that's a destructuring-bind. See also the pattern matching library which is already available in quicklisp.

**Comment by DruidGreeneyes** [comment]
Yes. What I'm hoping is that we can bind/alias one of them to a reader macro so that (among other things) they can be cleanly included in the parameter list of a function instead of taking up their own line:

    (defun foo ($(head . tail))
      ...)

or whatever characters you use to indicate the reader, should be equal to

    (defun foo (lis)
      (destructuring-bind (head . tail) 
          lis
        ...))

**Comment by noctuid** [comment]
@DruidGreeneyes You might find metabang-bind's [lambda-bind](https://common-lisp.net/project/metabang-bind/user-guide.html#header2-103) interesting if you aren't already familiar with it. Your `cl-dbind` looks nice as well.
